File: LeetCode/202快乐数.java
class Solution {
    public boolean isHappy(int n) {
        Set<Integer> set = new HashSet<>();
        while (n != 1 && !set.contains(n)) {
            set.add(n);
            n = calc(n);
        }
        return n == 1;
    }

    public int calc(int num) {
        int sum = 0;
        while (num > 0) {
            int temp = num % 10;
            sum += temp * temp;
            num /= 10;
        }
        return sum;
    }
}

File: LeetCode/202快乐数.go
func isHappy(n int) bool {
	set := map[int]bool{}
	for n != 1 && !set[n] {
		set[n] = true
		temp := func(x int) int {
			sum := 0
			for x > 0 {
				temp := x % 10
				sum += temp * temp
				x /= 10
			}
			return sum
		}
		n = temp(n)
	}
	return n == 1
}